sql/src/main/java/io/crate/jobs/PageDownstreamContext.java

```diff
@@ -95,10 +96,43 @@ public void setBucket(int bucketIdx, Bucket rows, boolean isLast, PageConsumeLis
             if (allFuturesSet.get(bucketIdx)) {
                 throw new IllegalStateException("May not set the same bucket of a page more than once");
             }
+            synchronized (listeners) {
+                listeners.add(pageConsumeListener);
+            }
 
             if (pageEmpty()) {
                 LOGGER.trace("calling nextPage");
-                pageDownstream.nextPage(new BucketPage(bucketFutures), pageConsumeListener);
+                pageDownstream.nextPage(new BucketPage(bucketFutures), new PageConsumeListener() {
+                    @Override
+                    public void needMore() {
+                        synchronized (listeners) {
+                        LOGGER.trace("calling needMore on all listeners({})", listeners.size());
+                            for (PageConsumeListener listener : listeners) {
+                                listener.needMore();
+                            }
+                            listeners.clear();
+                        }
+                    }
+
+                    @Override
+                    public void finish() {
+                        synchronized (listeners) {
+                        LOGGER.trace("calling finish() on all listeners({})", listeners.size());
+                            for (PageConsumeListener listener : listeners) {
+                                listener.finish();
+                            }
+                            listeners.clear();
+                        }
+                    }
+                });
+            }
+
+            // set all exhausted upstream futures
```

Comment:
because upstreams are not sending (useless) requests if they are exhausted, they are finished. but futures still exists (page buckets) and downstream is waiting for them